### PR TITLE
feat(discovery): add ability to hide classes from discovery

### LINKED ISSUE #338: Ability to skip any file from discovery

### PR DESCRIPTION
**Pull request by innocenzi**
This pull request adds the ability to hide classes from all discovery implementations by adding a `HideFromDiscovery` attribute (**edit**: updated to `DoNotDiscover`).

Happy to change the attribute name to anything else:
- `SkipDiscovery`
- `IgnoreFromDiscovery`
- `WithoutDiscovery`
- `\Tempest\Core\Discovery\Ignore`
- `\Tempest\Core\Discovery\Skip`

Closes #338
Related to #509

### CODE REVIEW

**Comment by brendt** [comment]
Nice, yeah the name, here are my preferences:

- `#[NoDiscovery]`
- `#[SkipDiscovery]`
- `#[DoNotDiscover]`

I think I like `#[DoNotDiscover]` the most, but I don't have a super strong opinion on it

**Comment by innocenzi** [comment]
I like it! I updated the pull request.

File: tests/Integration/Core/LoadDiscoveryClassesTest.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+        $this->kernel->discoveryLocations = [
+            realpath(__DIR__.'../../Fixtures/Discovery'),
+        ];
+
+        (new LoadDiscoveryClasses($this->kernel, $this->container));
+
+        $migrations = get(DatabaseConfig::class)->getMigrations();
+
+        assertNotContains(HiddenMigration::class, $migrations);
```
Comment (brendt) [inline tests/Integration/Core/LoadDiscoveryClassesTest.php:34]: Just a style preference, but I'd use $this->assertNotContains, but it's not a blocker

**Comment by brendt** [comment]
Nice!

**Comment by aidan-casey** [comment]
@brendt does it make sense to have this configurable too?

**Comment by brendt** [comment]
@aidan-casey configurable how?

**Comment by aidan-casey** [comment]
@brendt by adding to a config file. Otherwise, the developer has no real way to _not discover_ vendor files unless the package maintainer opts to not have them discovered.

**Comment by brendt** [comment]
Aaah gotcha. Yeah… that might make sense indeed 🤔

**Pull request merged**